### PR TITLE
Yeet `owning_ref`

### LINKED ISSUE #109974: `OwningRef` is included in the compiler despite known soundness problems.
The [`OwningRef` module in rustc_data_structures](https://doc.rust-lang.org/beta/nightly-rustc/rustc_data_structures/owning_ref/index.html), which seems to be a copy of [this crate](https://docs.rs/owning_ref/latest/owning_ref/index.html) (perhaps with small additions?), has known soundness issues.

I wrote an article about the soundness problems [here](https://github.com/noamtashma/owning-ref-unsoundness).

It's used in rustc in [`MetadataRef`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_data_structures/sync/type.MetadataRef.html) and related things. (I don't know if it's used anywhere else).

I don't know of any actual unsafe use of this crate in rustc, but I haven't actually tried to look for any.

IMO, in order to keep the code safe, either `owning_ref` should be replaced with something else, or otherwise, it should be updated to contain the known fixes to these soundness problems ([here](https://github.com/Kimundi/owning-ref-rs/pull/78)).

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Based on the discussions from https://github.com/rust-lang/rust/pull/109948

This replaces `owning_ref` with a far simpler & safer abstraction.

Fixes #109974

### CODE REVIEW

**Comment by petrochenkov** [comment]
Could someone summarize why this type is necessary?
If you have an owned `Box` or `Mmap`, why do you need to additionally pack a reference with it?

File: compiler/rustc_data_structures/src/owned_slice.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    try_slice_owned(owner, |x| Ok::<_, !>(slicer(x))).into_ok()
+}
+
+/// Makes an [`OwnedSlice`] out of an `owner` and a `slicer` function that can fail.
+///
+/// See [`slice_owned`] for the infallible version.
+pub fn try_slice_owned<O, F, E>(owner: O, slicer: F) -> Result<OwnedSlice, E>
+where
+    O: Send + Sync + 'static,
+    F: Fn(&O) -> Result<&[u8], E>,
```
Comment (Nilstrieb) [inline compiler/rustc_data_structures/src/owned_slice.rs:73]: `FnOnce`, this avoids having to change it in other places and is simply more correct.

**Review by Noratrieb — CHANGES_REQUESTED** [review]
It's a lovingly simple datastructure, but even lovingly simple datastructures deserve tests 😅

File: compiler/rustc_codegen_ssa/src/back/metadata.rs [hunk 2]
```diff
@@ -38,14 +37,14 @@ pub struct DefaultMetadataLoader;
...
     let file =
         File::open(path).map_err(|e| format!("failed to open file '{}': {}", path.display(), e))?;
-    let data = unsafe { Mmap::map(file) }
-        .map_err(|e| format!("failed to mmap file '{}': {}", path.display(), e))?;
-    let metadata = OwningRef::new(data).try_map(f)?;
-    return Ok(rustc_erase_owner!(metadata.map_owner_box()));
+
+    unsafe { Mmap::map(file) }
+        .map_err(|e| format!("failed to mmap file '{}': {}", path.display(), e))
+        .and_then(|mmap| try_slice_owned(mmap, |mmap| f(mmap)))
```
Comment (cramertj) [inline compiler/rustc_codegen_ssa/src/back/metadata.rs:47]: > `|mmap| f(mmap)`

nit: Can this be replaced with `f`?

Comment (WaffleLapkin) [reply compiler/rustc_codegen_ssa/src/back/metadata.rs:47]: No.

**Comment by noamtashma** [comment]
This isn't necessarily relevant to yeeting owning_ref, but this issue caused me to take a look at the nested `Box<dyn {Lrc<Box<dyn {?}>}>` structure in [def_path_hash_map](https://github.com/rust-lang/rust/pull/109971/commits/c0ceefdfaf9a1e131bfbf4849bbaaac23af621a4#diff-c3c088b211dfcbce123d78e915a564a3102d2665762f89c195c06e6920486947) and I think it could be improved further.

It seems that the metadata is only actually used in the form of `MetadataRef`, so that it's cloneable. If that's true, than we could instead always use the structure of `Lrc<dyn Send + Sync>`.

In other words, have the `owning_slice` contain an `Lrc<dyn Send + Sync>` instead of a `Box<dyn Send + Sync>`.
Then `owning_slice` becomes `clone`, it gains a `slice` method or equivalent to get a subslice of the existing slice without an extra indirection, and `MetadataRef` and `MetadataBlob` become equivalent, which is nice.

Any opinions?

**Comment by Nilstrieb** [comment]
try was cancelled by the push
@bors try @rust-timer queue

File: compiler/rustc_codegen_ssa/src/back/metadata.rs (hunk 2 above)
Comment (kadiwa4) [reply compiler/rustc_codegen_ssa/src/back/metadata.rs:47]: not with `&f` either?

Comment (WaffleLapkin) [reply compiler/rustc_codegen_ssa/src/back/metadata.rs:47]: No.

File: compiler/rustc_data_structures/src/owned_slice/tests.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+fn drop() {
+    let flag = Rc::new(Cell::new(false));
+    let flag_prime = Rc::clone(&flag);
+    let d = OnDrop(move || flag_prime.set(true));
+
+    let slice = slice_owned(d, |_| &[]);
+
+    assert_eq!(flag.get(), false);
+
+    std::mem::drop(slice);
```
Comment (Nilstrieb) [inline compiler/rustc_data_structures/src/owned_slice/tests.rs:59]:
```suggestion
    drop(slice);
```
`drop` is in the prelude.

**Review by Noratrieb — APPROVED** [review]
Have you run it through Miri? It looks very good but just to make sure.
r=me after the nit and Miri being green

Now with that, `rustc_data_structures` should be entirely Miri-green! We should look into asserting that in CI to avoid regressing it in the future. Could you open an issue for that afterwards?

File: compiler/rustc_data_structures/src/owned_slice/tests.rs (hunk 3 above)
Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/owned_slice/tests.rs:59]: I shadowed it in the test name 💀

**Comment by WaffleLapkin** [comment]
I've run `compiler/rustc_data_structures && cargo +nightly miri test` locally and it succeeded ✔️

**Comment by WaffleLapkin** [comment]
@Nilstrieb I've opened https://github.com/rust-lang/rust/issues/110015

**Comment by Nilstrieb** [comment]
Yeet!
@bors r+ rollup=maybe

File: compiler/rustc_data_structures/src/owned_slice.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    fn deref(&self) -> &[u8] {
+        // Safety:
+        // `self.bytes` is valid per the construction in `slice_owned`
+        // (which is the only constructor)
+        unsafe { &*self.bytes }
+    }
+}
+
+impl Borrow<[u8]> for OwnedSlice {
+    fn borrow(&self) -> &[u8] {
```
Comment (klensy) [inline compiler/rustc_data_structures/src/owned_slice.rs:104]: It will be nice to check, that this (and other short fns around) call will be inlined, as rust sometimes forget about that.

Comment (Nilstrieb) [reply compiler/rustc_data_structures/src/owned_slice.rs:104]: Great call, these two functions should all have `#[inline]` attributes. I'm not entirely sure where they are all used but given that it's used in metadata loading which is very hot this could be really bad on platforms that aren't compiled with LTO (like our neutral perf run was).
